src/test/memory.rs

```diff
@@ -58,81 +58,121 @@ where
 
     // Attempt conflicting write with wrong guard value
     let conflict_result = memory
         .guarded_write(
-            (T::Address::default(), None),
-            vec![(T::Address::default(), T::Word::default())],
+            (T::Address::from(random_address), None),
+            vec![(
+                T::Address::from(random_address),
+                T::Word::from(rng.gen::<u128>()),
+            )],
         )
         .await
         .unwrap();
 
     // Should return current value and not perform write
     assert_eq!(
         conflict_result,
-        Some(T::Word::default()),
-        "test_wrong_guard failed : {:?}",
-        conflict_result
+        Some(T::Word::from(word_to_write)),
+        "test_wrong_guard failed.\nExpected value {:?} after write. Got : {:?}.\nDebug seed : {:?}",
+        conflict_result,
+        Some(T::Word::from(word_to_write)),
+        seed
     );
 
     // Verify value wasn't changed
     let read_result = memory
-        .batch_read(vec![T::Address::default()])
+        .batch_read(vec![T::Address::from(random_address)])
         .await
         .unwrap();
-    assert_eq!(read_result, vec![Some(T::Word::default())]);
+    assert_eq!(
+        vec![Some(T::Word::from(word_to_write)),],
+        read_result,
+        "test_wrong_guard failed. Value was overwritten, violating the guard. Expected : {:?}, got : {:?}. Debug seed : {:?}",
+        vec![Some(T::Word::from(word_to_write)),],
+        read_result,
+        seed
+    );
 }
 
 pub async fn test_guarded_write_concurrent<T>(memory: T, seed: [u8; 32])
 where
-    T: MemoryADT<Address = u8, Word = u8> + Send + 'static + Clone,
+    T: MemoryADT<Address = u128, Word = u128> + Send + 'static + Clone,
     T::Error: std::error::Error,
 {
     {
         const N: usize = 1000; // number of threads
         let mut rng = StdRng::from_seed(seed);
-        let guard_address = rng.gen::<u8>(); // address used as guard
+        let counter_addr = rng.gen::<u128>(); // Random address for a counter
+
+        // Initialize counter to 0
+        memory
+            .guarded_write((counter_addr, None), vec![(counter_addr, 0)])
+            .await
+            .unwrap();
 
         let handles: Vec<_> = (0..N)
             .map(|_| {
                 let mem = memory.clone();
-                let adr = guard_address;
                 std::thread::spawn(move || async move {
-                    // All concurrent tasks will try to write to the same address
-                    // As the operation is atomic, only one of them should succeed and return true.
-                    mem.guarded_write((adr, None), vec![(guard_address, 1)])
-                        .await
-                        .unwrap()
-                        .is_none()
+                    loop {
+                        // Read current value
+                        let current = mem.batch_read(vec![counter_addr]).await.unwrap()[0]
+                            .expect("Counter should exist");
```

Comment:
No need for this: first try with `None`, then use the value returned by the previous `gwrite`.

Reply:
done